File: code/packages/elixir/pixel_container/test/pixel_container_test.exs
defmodule CodingAdventures.PixelContainerTest do
  use ExUnit.Case, async: true

  alias CodingAdventures.PixelContainer, as: PC

  # ─────────────────────────────────────────────────────────────────────────────
  # new/2
  # ─────────────────────────────────────────────────────────────────────────────

  describe "new/2" do
    test "creates a container with the correct width and height" do
      c = PC.new(10, 20)
      assert c.width == 10
      assert c.height == 20
    end

    test "buffer has correct byte size: width * height * 4" do
      c = PC.new(8, 6)
      assert byte_size(c.data) == 8 * 6 * 4
    end

    test "all bytes are zero initially (transparent black)" do
      c = PC.new(4, 4)
      # A binary of all zeros equals a binary filled with zero bytes
      assert c.data == :binary.copy(<<0, 0, 0, 0>>, 4 * 4)
    end

    test "1x1 container has 4 bytes" do
      c = PC.new(1, 1)
      assert byte_size(c.data) == 4
    end

    test "0x0 container has 0 bytes (degenerate but valid)" do
      c = PC.new(0, 0)
      assert byte_size(c.data) == 0
    end

    test "large container has correct size" do
      c = PC.new(1920, 1080)
      assert byte_size(c.data) == 1920 * 1080 * 4
    end
  end

  # ─────────────────────────────────────────────────────────────────────────────
  # pixel_at/3
  # ─────────────────────────────────────────────────────────────────────────────

  describe "pixel_at/3" do
    test "all pixels in a new container are {0, 0, 0, 0}" do
      c = PC.new(5, 5)

      for x <- 0..4, y <- 0..4 do
        assert PC.pixel_at(c, x, y) == {0, 0, 0, 0},
               "expected {0,0,0,0} at (#{x}, #{y})"
      end
    end

    test "returns {0, 0, 0, 0} for negative x" do
      c = PC.new(5, 5)
      assert PC.pixel_at(c, -1, 0) == {0, 0, 0, 0}
    end

    test "returns {0, 0, 0, 0} for negative y" do
      c = PC.new(5, 5)
      assert PC.pixel_at(c, 0, -1) == {0, 0, 0, 0}
    end

    test "returns {0, 0, 0, 0} for x equal to width (off-by-one boundary)" do
      c = PC.new(5, 5)
      assert PC.pixel_at(c, 5, 0) == {0, 0, 0, 0}
    end

    test "returns {0, 0, 0, 0} for y equal to height (off-by-one boundary)" do
      c = PC.new(5, 5)
      assert PC.pixel_at(c, 0, 5) == {0, 0, 0, 0}
    end

    test "returns {0, 0, 0, 0} for x far out of bounds" do
      c = PC.new(5, 5)
      assert PC.pixel_at(c, 100, 100) == {0, 0, 0, 0}
    end
  end

  # ─────────────────────────────────────────────────────────────────────────────
  # set_pixel/7
  # ─────────────────────────────────────────────────────────────────────────────

  describe "set_pixel/7" do
    test "sets pixel at (0, 0) correctly" do
      c = PC.new(5, 5)
      c = PC.set_pixel(c, 0, 0, 255, 128, 64, 255)
      assert PC.pixel_at(c, 0, 0) == {255, 128, 64, 255}
    end

    test "sets pixel at bottom-right corner correctly" do
      c = PC.new(5, 5)
      c = PC.set_pixel(c, 4, 4, 1, 2, 3, 4)
      assert PC.pixel_at(c, 4, 4) == {1, 2, 3, 4}
    end

    test "setting one pixel does not affect neighbors" do
      c = PC.new(5, 5)
      c = PC.set_pixel(c, 2, 2, 200, 100, 50, 255)
      # Immediate neighbors should remain untouched
      assert PC.pixel_at(c, 1, 2) == {0, 0, 0, 0}
      assert PC.pixel_at(c, 3, 2) == {0, 0, 0, 0}
      assert PC.pixel_at(c, 2, 1) == {0, 0, 0, 0}
      assert PC.pixel_at(c, 2, 3) == {0, 0, 0, 0}
    end

    test "can overwrite a pixel multiple times" do
      c = PC.new(3, 3)
      c = PC.set_pixel(c, 1, 1, 10, 20, 30, 40)
      assert PC.pixel_at(c, 1, 1) == {10, 20, 30, 40}
      c = PC.set_pixel(c, 1, 1, 99, 88, 77, 66)
      assert PC.pixel_at(c, 1, 1) == {99, 88, 77, 66}
    end

    test "out-of-bounds set returns original container unchanged" do
      c = PC.new(3, 3)
      c2 = PC.set_pixel(c, 100, 100, 255, 0, 0, 255)
      assert c2 == c
    end

    test "negative x is no-op" do
      c = PC.new(3, 3)
      c2 = PC.set_pixel(c, -1, 0, 255, 0, 0, 255)
      assert c2 == c
    end

    test "x == width is no-op (off-by-one guard)" do
      c = PC.new(3, 3)
      c2 = PC.set_pixel(c, 3, 0, 255, 0, 0, 255)
      assert c2 == c
    end

    test "can set all 4 channels to max (255)" do
      c = PC.new(2, 2)
      c = PC.set_pixel(c, 0, 0, 255, 255, 255, 255)
      assert PC.pixel_at(c, 0, 0) == {255, 255, 255, 255}
    end

    test "can set all 4 channels to zero (transparent black)" do
      c = PC.new(2, 2)
      c = PC.set_pixel(c, 0, 0, 255, 255, 255, 255)
      c = PC.set_pixel(c, 0, 0, 0, 0, 0, 0)
      assert PC.pixel_at(c, 0, 0) == {0, 0, 0, 0}
    end
  end

  # ─────────────────────────────────────────────────────────────────────────────
  # fill_pixels/5
  # ─────────────────────────────────────────────────────────────────────────────

  describe "fill_pixels/5" do
    test "fills all pixels with the given color" do
      c = PC.new(4, 4)
      c = PC.fill_pixels(c, 255, 0, 128, 200)

      for x <- 0..3, y <- 0..3 do
        assert PC.pixel_at(c, x, y) == {255, 0, 128, 200},
               "expected fill color at (#{x}, #{y})"
      end
    end

    test "fill with white produces all-255 data" do
      c = PC.new(3, 3)
      c = PC.fill_pixels(c, 255, 255, 255, 255)
      assert c.data == :binary.copy(<<255, 255, 255, 255>>, 9)
    end

    test "fill with transparent black produces all-zero data" do
      c = PC.new(3, 3)
      c = PC.fill_pixels(c, 0, 0, 0, 0)
      assert c.data == :binary.copy(<<0, 0, 0, 0>>, 9)
    end

    test "fill preserves width and height" do
      c = PC.new(7, 5)
      c = PC.fill_pixels(c, 100, 100, 100, 255)
      assert c.width == 7
      assert c.height == 5
    end

    test "fill overwrites previously set pixels" do
      c = PC.new(3, 3)
      c = PC.set_pixel(c, 1, 1, 255, 0, 0, 255)
      c = PC.fill_pixels(c, 0, 255, 0, 255)
      assert PC.pixel_at(c, 1, 1) == {0, 255, 0, 255}
    end
  end

  # ─────────────────────────────────────────────────────────────────────────────
  # byte_size/1
  # ─────────────────────────────────────────────────────────────────────────────

  describe "byte_size/1" do
    test "returns width * height * 4" do
      c = PC.new(6, 7)
      assert PC.byte_size(c) == 6 * 7 * 4
    end

    test "matches byte_size of the data binary directly" do
      c = PC.new(10, 10)
      assert PC.byte_size(c) == byte_size(c.data)
    end
  end

  # ─────────────────────────────────────────────────────────────────────────────
  # Round-trip: set then read many pixels
  # ─────────────────────────────────────────────────────────────────────────────

  describe "round-trip set/get" do
    test "can independently set and read every pixel in a 4x4 grid" do
      c = PC.new(4, 4)

      # Use (x + y * 16) as a unique value per pixel for each channel offset
      c =
        for x <- 0..3, y <- 0..3, reduce: c do
          acc ->
            r = rem(x * 17 + y * 31, 256)
            g = rem(x * 53 + y * 7, 256)
            b = rem(x * 97 + y * 13, 256)
            a = 255
            PC.set_pixel(acc, x, y, r, g, b, a)
        end

      for x <- 0..3, y <- 0..3 do
        r = rem(x * 17 + y * 31, 256)
        g = rem(x * 53 + y * 7, 256)
        b = rem(x * 97 + y * 13, 256)
        assert PC.pixel_at(c, x, y) == {r, g, b, 255}
      end
    end
  end
end

File: code/packages/go/pixel-container/pixel_container_test.go
// Package pixelcontainer_test exercises every exported function in the
// pixelcontainer package.
//
// Tests are organised in three groups:
//  1. Construction — New() creates valid, zero-filled containers.
//  2. PixelAt / SetPixel — reads and writes including out-of-bounds behaviour.
//  3. FillPixels — bulk-fill and dimension checks.
package pixelcontainer_test

import (
	"testing"

	pc "github.com/adhithyan15/coding-adventures/code/packages/go/pixel-container"
)

// ── Construction ──────────────────────────────────────────────────────────────

// TestNew_Dimensions checks that Width and Height are stored correctly.
func TestNew_Dimensions(t *testing.T) {
	img := pc.New(10, 20)
	if img.Width != 10 {
		t.Errorf("Width: got %d, want 10", img.Width)
	}
	if img.Height != 20 {
		t.Errorf("Height: got %d, want 20", img.Height)
	}
}

// TestNew_DataLength verifies that the backing slice is exactly W*H*4 bytes.
func TestNew_DataLength(t *testing.T) {
	img := pc.New(7, 3)
	want := 7 * 3 * 4
	if len(img.Data) != want {
		t.Errorf("len(Data): got %d, want %d", len(img.Data), want)
	}
}

// TestNew_ZeroFilled checks that all bytes start at zero.
func TestNew_ZeroFilled(t *testing.T) {
	img := pc.New(4, 4)
	for i, b := range img.Data {
		if b != 0 {
			t.Errorf("Data[%d] = %d, want 0", i, b)
		}
	}
}

// TestNew_1x1 checks a single-pixel image.
func TestNew_1x1(t *testing.T) {
	img := pc.New(1, 1)
	if len(img.Data) != 4 {
		t.Errorf("1×1 image should have 4 bytes, got %d", len(img.Data))
	}
}

// TestNew_LargeImage ensures no panic on a large allocation.
func TestNew_LargeImage(t *testing.T) {
	img := pc.New(1920, 1080)
	want := uint32(1920)
	if img.Width != want {
		t.Errorf("Width: got %d, want %d", img.Width, want)
	}
	if uint64(len(img.Data)) != uint64(1920)*uint64(1080)*4 {
		t.Error("Data length mismatch for 1920x1080 image")
	}
}

// ── PixelAt ──────────────────────────────────────────────────────────────────

// TestPixelAt_ZeroAfterNew confirms all pixels read as (0,0,0,0) after New.
func TestPixelAt_ZeroAfterNew(t *testing.T) {
	img := pc.New(5, 5)
	r, g, b, a := pc.PixelAt(img, 2, 3)
	if r != 0 || g != 0 || b != 0 || a != 0 {
		t.Errorf("PixelAt on new image: got (%d,%d,%d,%d), want (0,0,0,0)", r, g, b, a)
	}
}

// TestPixelAt_OutOfBoundsX checks that reading past the right edge returns zeros.
func TestPixelAt_OutOfBoundsX(t *testing.T) {
	img := pc.New(4, 4)
	r, g, b, a := pc.PixelAt(img, 4, 0) // x == Width is out of bounds
	if r != 0 || g != 0 || b != 0 || a != 0 {
		t.Errorf("out-of-bounds x: got (%d,%d,%d,%d), want (0,0,0,0)", r, g, b, a)
	}
}

// TestPixelAt_OutOfBoundsY checks that reading past the bottom edge returns zeros.
func TestPixelAt_OutOfBoundsY(t *testing.T) {
	img := pc.New(4, 4)
	r, g, b, a := pc.PixelAt(img, 0, 4) // y == Height is out of bounds
	if r != 0 || g != 0 || b != 0 || a != 0 {
		t.Errorf("out-of-bounds y: got (%d,%d,%d,%d), want (0,0,0,0)", r, g, b, a)
	}
}

// TestPixelAt_FarOutOfBounds uses large coordinates to confirm no panic.
func TestPixelAt_FarOutOfBounds(t *testing.T) {
	img := pc.New(4, 4)
	r, g, b, a := pc.PixelAt(img, 9999, 9999)
	if r != 0 || g != 0 || b != 0 || a != 0 {
		t.Errorf("far out-of-bounds: got (%d,%d,%d,%d), want (0,0,0,0)", r, g, b, a)
	}
}

// ── SetPixel ─────────────────────────────────────────────────────────────────

// TestSetPixel_Basic writes a pixel and reads it back.
func TestSetPixel_Basic(t *testing.T) {
	img := pc.New(8, 8)
	pc.SetPixel(img, 3, 2, 10, 20, 30, 255)
	r, g, b, a := pc.PixelAt(img, 3, 2)
	if r != 10 || g != 20 || b != 30 || a != 255 {
		t.Errorf("SetPixel/PixelAt: got (%d,%d,%d,%d), want (10,20,30,255)", r, g, b, a)
	}
}

// TestSetPixel_AllChannels verifies each channel is stored independently.
func TestSetPixel_AllChannels(t *testing.T) {
	img := pc.New(3, 3)
	pc.SetPixel(img, 0, 0, 1, 2, 3, 4)
	r, g, b, a := pc.PixelAt(img, 0, 0)
	if r != 1 {
		t.Errorf("R: got %d, want 1", r)
	}
	if g != 2 {
		t.Errorf("G: got %d, want 2", g)
	}
	if b != 3 {
		t.Errorf("B: got %d, want 3", b)
	}
	if a != 4 {
		t.Errorf("A: got %d, want 4", a)
	}
}

// TestSetPixel_MaxValues checks 255 in every channel (maximum RGBA values).
func TestSetPixel_MaxValues(t *testing.T) {
	img := pc.New(2, 2)
	pc.SetPixel(img, 1, 1, 255, 255, 255, 255)
	r, g, b, a := pc.PixelAt(img, 1, 1)
	if r != 255 || g != 255 || b != 255 || a != 255 {
		t.Errorf("max values: got (%d,%d,%d,%d), want (255,255,255,255)", r, g, b, a)
	}
}

// TestSetPixel_DoesNotClobberNeighbour sets one pixel and confirms the
// adjacent pixel is unaffected.
func TestSetPixel_DoesNotClobberNeighbour(t *testing.T) {
	img := pc.New(4, 4)
	pc.SetPixel(img, 1, 1, 100, 100, 100, 100)
	r, g, b, a := pc.PixelAt(img, 2, 1) // right neighbour
	if r != 0 || g != 0 || b != 0 || a != 0 {
		t.Errorf("neighbour contaminated: got (%d,%d,%d,%d), want (0,0,0,0)", r, g, b, a)
	}
}

// TestSetPixel_OutOfBoundsX ensures no panic and no write for x == Width.
func TestSetPixel_OutOfBoundsX(t *testing.T) {
	img := pc.New(4, 4)
	// Should be a no-op — must not panic.
	pc.SetPixel(img, 4, 0, 99, 99, 99, 99)
	// Verify data is still all zeros.
	for _, b := range img.Data {
		if b != 0 {
			t.Error("out-of-bounds SetPixel wrote to Data")
			break
		}
	}
}

// TestSetPixel_OutOfBoundsY ensures no panic and no write for y == Height.
func TestSetPixel_OutOfBoundsY(t *testing.T) {
	img := pc.New(4, 4)
	pc.SetPixel(img, 0, 4, 99, 99, 99, 99)
	for _, b := range img.Data {
		if b != 0 {
			t.Error("out-of-bounds SetPixel wrote to Data")
			break
		}
	}
}

// TestSetPixel_OverwriteExisting verifies a pixel can be updated.
func TestSetPixel_OverwriteExisting(t *testing.T) {
	img := pc.New(3, 3)
	pc.SetPixel(img, 0, 0, 10, 20, 30, 40)
	pc.SetPixel(img, 0, 0, 50, 60, 70, 80) // overwrite
	r, g, b, a := pc.PixelAt(img, 0, 0)
	if r != 50 || g != 60 || b != 70 || a != 80 {
		t.Errorf("overwrite: got (%d,%d,%d,%d), want (50,60,70,80)", r, g, b, a)
	}
}

// TestSetPixel_BottomRightCorner exercises the last pixel in the buffer.
func TestSetPixel_BottomRightCorner(t *testing.T) {
	img := pc.New(5, 5)
	pc.SetPixel(img, 4, 4, 11, 22, 33, 44)
	r, g, b, a := pc.PixelAt(img, 4, 4)
	if r != 11 || g != 22 || b != 33 || a != 44 {
		t.Errorf("bottom-right corner: got (%d,%d,%d,%d), want (11,22,33,44)", r, g, b, a)
	}
}

// ── FillPixels ───────────────────────────────────────────────────────────────

// TestFillPixels_White fills with white and checks every pixel.
func TestFillPixels_White(t *testing.T) {
	img := pc.New(3, 3)
	pc.FillPixels(img, 255, 255, 255, 255)
	for y := uint32(0); y < img.Height; y++ {
		for x := uint32(0); x < img.Width; x++ {
			r, g, b, a := pc.PixelAt(img, x, y)
			if r != 255 || g != 255 || b != 255 || a != 255 {
				t.Errorf("pixel (%d,%d): got (%d,%d,%d,%d), want (255,255,255,255)", x, y, r, g, b, a)
			}
		}
	}
}

// TestFillPixels_Transparent fills with transparent black.
func TestFillPixels_Transparent(t *testing.T) {
	img := pc.New(2, 2)
	// First paint something
	pc.FillPixels(img, 100, 100, 100, 100)
	// Then reset to transparent black
	pc.FillPixels(img, 0, 0, 0, 0)
	for _, b := range img.Data {
		if b != 0 {
			t.Errorf("expected all zeros after fill(0,0,0,0), got %d", b)
			break
		}
	}
}

// TestFillPixels_DistinctChannels fills with unequal channel values and
// checks every pixel to make sure no channel bleeds into another.
func TestFillPixels_DistinctChannels(t *testing.T) {
	img := pc.New(4, 3)
	pc.FillPixels(img, 11, 22, 33, 44)
	for y := uint32(0); y < img.Height; y++ {
		for x := uint32(0); x < img.Width; x++ {
			r, g, b, a := pc.PixelAt(img, x, y)
			if r != 11 || g != 22 || b != 33 || a != 44 {
				t.Errorf("pixel (%d,%d): got (%d,%d,%d,%d), want (11,22,33,44)", x, y, r, g, b, a)
			}
		}
	}
}

// ── Validate ─────────────────────────────────────────────────────────────────

// TestValidate_Valid checks that a properly constructed container passes.
func TestValidate_Valid(t *testing.T) {
	img := pc.New(10, 10)
	if err := pc.Validate(img); err != nil {
		t.Errorf("Validate on valid container: %v", err)
	}
}

// TestValidate_TruncatedData checks that a container with a short Data slice fails.
func TestValidate_TruncatedData(t *testing.T) {
	img := pc.New(4, 4)
	img.Data = img.Data[:10] // deliberately wrong length
	if err := pc.Validate(img); err == nil {
		t.Error("Validate should have returned an error for truncated Data")
	}
}
